clients/da-vinci-client/src/main/java/com/linkedin/davinci/validation/KafkaDataIntegrityValidator.java

```diff
@@ -62,12 +119,41 @@ public void validateMessage(
 
   /**
    * For a given partition, find all the producers that has written to this partition and update the offsetRecord using
-   * segment information.
-   * @param partition
-   * @param offsetRecord
+   * segment information. Prior to this, the state which is expired according to {@link #maxAgeInMs} will be cleared.
+   *
+   * @param partition to extract info for
+   * @param offsetRecord to modify
    */
   public void updateOffsetRecordForPartition(int partition, OffsetRecord offsetRecord) {
-    producerTrackerMap.values().forEach(producerTracker -> producerTracker.updateOffsetRecord(partition, offsetRecord));
+    if (this.maxAgeInMs != DISABLED) {
+      clearExpiredStateAndUpdateOffsetRecordForPartition(partition, offsetRecord);
+    } else {
+      Iterator<ProducerTracker> iterator = this.producerTrackerMap.values().iterator();
+      while (iterator.hasNext()) {
+        iterator.next().updateOffsetRecord(partition, offsetRecord);
+      }
+    }
+  }
+
+  void clearExpiredStateAndUpdateOffsetRecordForPartition(int partition, OffsetRecord offsetRecord) {
+    long minimumRequiredRecordProducerTimestamp = offsetRecord.getMaxMessageTimeInMs() - this.maxAgeInMs;
+    int numberOfClearedGUIDs = 0;
+    Iterator<ProducerTracker> iterator = this.producerTrackerMap.values().iterator();
+    ProducerTracker producerTracker;
+    while (iterator.hasNext()) {
+      producerTracker = iterator.next();
+      if (producerTracker.clearExpiredState(partition, minimumRequiredRecordProducerTimestamp)
+          && producerTracker.isEmpty()) {
```

Comment:
It seems there is a race condition here.
`ProducerTracker` is maintaining a state per partition and multiple partitions can update `ProducerTracker` concurrently via `validateMessage`.

Maybe we need to use some read-write lock to guard this logic.
Such as the clear logic will grab the write lock and all other operations need to grab the read lock.

Reply:
Hmm... I'll see what I can do to tweak this. Also, now that I look at this, having the `&& producerTracker.isEmpty()` as part of the same if condition might be wrong... maybe the ideal flow would be along the lines of:

```
if (producerTracker.clearExpiredState(partition, minimumRequiredRecordProducerTimestamp)) {
  // lock
  if (producerTracker.isEmpty()) {
    // remove state
  }
  // unlock
} else {
  producerTracker.updateOffsetRecord(partition, offsetRecord);
}
```

Because the else block should happen when the state of that specific partition is not expired, regardless of the state of the other partitions.